// kuzzle/lib/api/openapi/components/document/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiDocumentmCreateOrReplaceComponent = exports.OpenApiDocumentmCreateOrReplace = exports.OpenApiDocumentValidateComponent = exports.OpenApiDocumentValidate = exports.OpenApiDocumentDeleteByQueryComponent = exports.OpenApiDocumentDeleteByQuery = exports.OpenApiDocumentDeleteComponent = exports.OpenApiDocumentDelete = exports.OpenApiDocumentScrollComponent = exports.OpenApiDocumentScroll = exports.OpenApiDocumentUpdateComponent = exports.OpenApiDocumentUpdate = exports.OpenApiDocumentExistsComponent = exports.OpenApiDocumentExists = exports.OpenApiDocumentReplaceComponent = exports.OpenApiDocumentReplace = exports.OpenApiDocumentGetComponent = exports.OpenApiDocumentGet = exports.OpenApiDocumentCreateOrReplaceComponent = exports.OpenApiDocumentCreateOrReplace = exports.OpenApiDocumentCreateComponent = exports.OpenApiDocumentCreate = exports.OpenApiDocumentCountComponent = exports.OpenApiDocumentCount = void 0;
const readYamlFile_1 = require("../../../../util/readYamlFile");
// reading the description of the Count action in the controller document.
// The yaml objects are then stored in the variables below
const countObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/count.yaml");
exports.OpenApiDocumentCount = countObject.DocumentCount;
exports.OpenApiDocumentCountComponent = countObject.components.schemas;
// reading the description of the Create action in the controller document.
// The yaml objects are then stored in the variables below
const createObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/create.yaml");
exports.OpenApiDocumentCreate = createObject.DocumentCreate;
exports.OpenApiDocumentCreateComponent = createObject.components.schemas;
// reading the description of the CreateOrReplace action in the controller document.
// The yaml objects are then stored in the variables below
const createOrReplaceObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/createOrReplace.yaml");
exports.OpenApiDocumentCreateOrReplace = createOrReplaceObject.DocumentCreateOrReplace;
exports.OpenApiDocumentCreateOrReplaceComponent = createOrReplaceObject.components.schemas;
// reading the description of the Get action in the controller document.
// The yaml objects are then stored in the variables below
const getObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/get.yaml");
exports.OpenApiDocumentGet = getObject.DocumentGet;
exports.OpenApiDocumentGetComponent = getObject.components.schemas;
// reading the description of the Replace action in the controller document.
// The yaml objects are then stored in the variables below
const replaceObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/replace.yaml");
exports.OpenApiDocumentReplace = replaceObject.DocumentReplace;
exports.OpenApiDocumentReplaceComponent = replaceObject.components.schemas;
// reading the description of the Exists action in the controller document.
// The yaml objects are then stored in the variables below
const existsObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/exists.yaml");
exports.OpenApiDocumentExists = existsObject.DocumentExists;
exports.OpenApiDocumentExistsComponent = existsObject.components.schemas;
// reading the description of the Update action in the controller document.
// The yaml objects are then stored in the variables below
const updateObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/update.yaml");
exports.OpenApiDocumentUpdate = updateObject.DocumentUpdate;
exports.OpenApiDocumentUpdateComponent = updateObject.components.schemas;
// reading the description of the Scroll action in the controller document.
// The yaml objects are then stored in the variables below
const scrollObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/scroll.yaml");
exports.OpenApiDocumentScroll = scrollObject.DocumentScroll;
exports.OpenApiDocumentScrollComponent = scrollObject.components.schemas;
// reading the description of the Delete action in the controller document.
// The yaml objects are then stored in the variables below
const deleteObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/delete.yaml");
exports.OpenApiDocumentDelete = deleteObject.DocumentDelete;
exports.OpenApiDocumentDeleteComponent = deleteObject.components.schemas;
// reading the description of the DeleteByQuery action in the controller document.
// The yaml objects are then stored in the variables below
const deleteByQueryObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/deleteByQuery.yaml");
exports.OpenApiDocumentDeleteByQuery = deleteByQueryObject.DocumentDeleteByQuery;
exports.OpenApiDocumentDeleteByQueryComponent = deleteByQueryObject.components.schemas;
// reading the description of the Validate action in the controller document.
// The yaml objects are then stored in the variables below
const validateObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/validate.yaml");
exports.OpenApiDocumentValidate = validateObject.DocumentValidate;
exports.OpenApiDocumentValidateComponent = validateObject.components.schemas;
// reading the description of the mCreateOrReplace action in the controller document.
// The yaml objects are then stored in the variables below
const mCreateOrReplaceObject = (0, readYamlFile_1.readYamlFile)(__dirname + "/mCreateOrReplace.yaml");
exports.OpenApiDocumentmCreateOrReplace = mCreateOrReplaceObject.DocumentmCreateOrReplace;
exports.OpenApiDocumentmCreateOrReplaceComponent = mCreateOrReplaceObject.components.schemas;
