// kuzzle/lib/kerror/errors/kuzzleError.js
"use strict";
/*
 * Kuzzle, a backend software, self-hostable and ready to use
 * to power modern apps
 *
 * Copyright 2015-2022 Kuzzle
 * mailto: support AT kuzzle.io
 * website: http://kuzzle.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KuzzleError = void 0;
const util = __importStar(require("util"));
/**
 * API error are instances of this class.
 * See https://docs.kuzzle.io/core/2/api/errors/types/
 */
class KuzzleError extends Error {
    constructor(message, status, id, code) {
        super(message);
        this.status = status;
        this.code = code;
        this.id = id;
        this.props = undefined;
        this.stack = undefined;
        if (util.types.isNativeError(message)) {
            this.message = message.message;
            this.stack = message.stack;
        }
        else {
            this.message = message;
            Error.captureStackTrace(this, KuzzleError);
        }
    }
    /**
     * Error class name (e.g: 'NotFoundError')
     */
    get name() {
        return this.constructor.name;
    }
    toJSON() {
        return {
            code: this.code,
            id: this.id,
            message: this.message,
            props: this.props,
            stack: this.stack,
            status: this.status,
        };
    }
}
exports.KuzzleError = KuzzleError;
